src/common/common.h

```diff
@@ -78,3 +81,36 @@ std::wstring get_resource_string(UINT resource_id, HINSTANCE instance, const wch
 //  extern "C" IMAGE_DOS_HEADER __ImageBase;
 // is added to the .cpp file.
 #define GET_RESOURCE_STRING(resource_id) get_resource_string(resource_id, reinterpret_cast<HINSTANCE>(&__ImageBase), L#resource_id)
+
+// Helper class for various COM-related APIs, e.g working with security descriptors
+template<typename T>
+struct typed_storage
+{
+    std::unique_ptr<char[]> _buffer;
+    inline explicit typed_storage(const DWORD size) :
+        _buffer{ std::make_unique<char[]>(size) }
+    {
+    }
+    inline operator T*()
+    {
+        return reinterpret_cast<T*>(_buffer.get());
+    }
+};
+
+template<typename Callable>
+struct on_scope_exit_impl
+{
+    Callable _f;
+    on_scope_exit_impl(Callable f) :
+        _f{ std::move(f) } {}
+
+    ~on_scope_exit_impl()
+    {
+        _f();
+    }
+};
+
+#define ON_SCOPE_EXIT_CONCAT(a, b) ON_SCOPE_EXIT_XCONCAT(a, b)
+#define ON_SCOPE_EXIT_XCONCAT(a, b) a##b
+#define ON_SCOPE_EXIT(func) \
```

Comment:
I think we should have an explicit object and not use a macro. I.e. in the `main.cpp`:

```c++
on_scope_exit free_realtive_sd{[&]{ LocalFree(self_relative_sd); });
```

Reply:
(or, if we want a macro, we can just put `func();` in the destructor and not do the move thingy)

Reply:
It always feels good to delete macros 😋